File: vsub/video.py
"""视频处理模块"""

import json
import shutil
import subprocess
from functools import lru_cache
from pathlib import Path
from typing import Optional


class VideoInfo:
    """视频信息"""

    def __init__(
        self,
        path: Path,
        duration: float = 0.0,
        width: int = 0,
        height: int = 0,
        has_audio: bool = False,
        audio_codec: Optional[str] = None,
        video_codec: Optional[str] = None,
        fps: float = 0.0,
    ):
        self.path = path
        self.duration = duration
        self.width = width
        self.height = height
        self.has_audio = has_audio
        self.audio_codec = audio_codec
        self.video_codec = video_codec
        self.fps = fps

    @property
    def resolution(self) -> str:
        return f"{self.width}x{self.height}"

    def __repr__(self) -> str:
        return (
            f"VideoInfo({self.path.name}, {self.resolution}, "
            f"{self.fps:.2f}fps, {self.duration:.2f}s, audio={self.has_audio})"
        )


@lru_cache(maxsize=1)
def check_ffmpeg() -> Path:
    """检查 FFmpeg 是否可用（带缓存）"""
    ffmpeg_path = shutil.which("ffmpeg")
    if not ffmpeg_path:
        raise RuntimeError("未找到 FFmpeg，请确保 FFmpeg 已安装并在 PATH 中")
    return Path(ffmpeg_path)


@lru_cache(maxsize=1)
def check_ffprobe() -> Path:
    """检查 ffprobe 是否可用（带缓存）"""
    ffprobe_path = shutil.which("ffprobe")
    if not ffprobe_path:
        raise RuntimeError("未找到 ffprobe，请确保 FFmpeg 已安装并在 PATH 中")
    return Path(ffprobe_path)


def probe_video(path: Path) -> VideoInfo:
    """使用 ffprobe 探测视频信息"""
    ffprobe = check_ffprobe()

    cmd = [
        str(ffprobe),
        "-v", "error",
        "-print_format", "json",
        "-show_streams",
        "-show_format",
        str(path),
    ]

    result = subprocess.run(cmd, capture_output=True, text=True)
    if result.returncode != 0:
        # 清理错误消息，移除敏感路径信息
        error_msg = _sanitize_ffprobe_error(result.stderr)
        raise RuntimeError(f"ffprobe 失败: {error_msg}")

    try:
        data = json.loads(result.stdout)
    except json.JSONDecodeError as e:
        raise RuntimeError(f"解析 ffprobe 输出失败: {e}")

    return parse_video_info(path, data)


def _sanitize_ffprobe_error(error_msg: str) -> str:
    """清理 ffprobe 错误消息中的敏感信息"""
    lines = error_msg.split('\n')
    sanitized = []
    for line in lines[:3]:  # 只保留前3行
        # 如果行太长，截断
        if len(line) > 150:
            line = line[:150] + "..."
        sanitized.append(line)
    return '\n'.join(sanitized)


def parse_video_info(path: Path, data: dict) -> VideoInfo:
    """解析 ffprobe 输出"""
    streams = data.get("streams", [])
    format_info = data.get("format", {})

    # 查找视频流
    video_stream = None
    audio_stream = None

    for stream in streams:
        codec_type = stream.get("codec_type")
        if codec_type == "video":
            video_stream = stream
        elif codec_type == "audio":
            audio_stream = stream

    # 获取时长
    duration = 0.0
    if "duration" in format_info:
        duration = float(format_info["duration"])
    elif video_stream and "duration" in video_stream:
        duration = float(video_stream["duration"])

    # 获取分辨率
    width = 0
    height = 0
    fps = 0.0
    video_codec = None

    if video_stream:
        width = video_stream.get("width", 0)
        height = video_stream.get("height", 0)
        video_codec = video_stream.get("codec_name")

        # 解析帧率（使用 float 避免溢出）
        r_frame_rate = video_stream.get("r_frame_rate", "")
        if "/" in r_frame_rate:
            try:
                num, den = r_frame_rate.split("/")
                den_val = float(den)
                if den_val > 0:
                    fps = float(num) / den_val
            except (ValueError, ZeroDivisionError):
                fps = 0.0

    # 获取音频信息
    audio_codec = None
    has_audio = audio_stream is not None
    if audio_stream:
        audio_codec = audio_stream.get("codec_name")

    return VideoInfo(
        path=path,
        duration=duration,
        width=width,
        height=height,
        has_audio=has_audio,
        audio_codec=audio_codec,
        video_codec=video_codec,
        fps=fps,
    )


def validate_video(path: Path) -> None:
    """验证视频文件"""
    if not path.exists():
        raise FileNotFoundError(f"文件不存在: {path}")

    if not path.is_file():
        raise ValueError(f"不是文件: {path}")

    # 使用 ffprobe 验证
    probe_video(path)

File: src/core/video.rs
use std::path::Path;
use std::process::Command;
use tracing::{debug, trace};

use crate::VsubError;

/// 视频信息
#[derive(Debug, Clone)]
pub struct VideoInfo {
    pub path: std::path::PathBuf,
    pub duration: f64,
    pub width: u32,
    pub height: u32,
    pub has_audio: bool,
    pub audio_codec: Option<String>,
    pub video_codec: Option<String>,
    pub fps: f64,
}

impl VideoInfo {
    pub fn duration_seconds(&self) -> f64 {
        self.duration
    }

    pub fn resolution(&self) -> String {
        format!("{}x{}", self.width, self.height)
    }
}

/// 检查 FFmpeg 是否可用
pub fn check_ffmpeg() -> Result<std::path::PathBuf, VsubError> {
    let output = Command::new("which")
        .arg("ffmpeg")
        .output()
        .map_err(|e| VsubError::Io(e))?;

    if !output.status.success() {
        return Err(VsubError::FFmpegNotFound);
    }

    let path = String::from_utf8_lossy(&output.stdout).trim().to_string();
    Ok(std::path::PathBuf::from(path))
}

/// 使用 ffprobe 探测视频信息
pub fn probe_video<P: AsRef<Path>>(path: P) -> Result<VideoInfo, VsubError> {
    let path = path.as_ref();

    trace!("Probing video: {}", path.display());

    let output = Command::new("ffprobe")
        .args(&[
            "-v", "error",
            "-print_format", "json",
            "-show_streams",
            "-show_format",
            path.to_str().unwrap(),
        ])
        .output()
        .map_err(|e| {
            if e.kind() == std::io::ErrorKind::NotFound {
                VsubError::FFmpegNotFound
            } else {
                VsubError::Io(e)
            }
        })?;

    if !output.status.success() {
        let stderr = String::from_utf8_lossy(&output.stderr);
        return Err(VsubError::Video(format!("ffprobe 失败: {}", stderr)));
    }

    let json: serde_json::Value = serde_json::from_slice(&output.stdout)
        .map_err(|e| VsubError::Video(format!("解析 ffprobe 输出失败: {}", e)))?;

    parse_video_info(path, &json)
}

fn parse_video_info(path: &Path, json: &serde_json::Value) -> Result<VideoInfo, VsubError> {
    let streams = json.get("streams")
        .and_then(|s| s.as_array())
        .ok_or_else(|| VsubError::Video("无法获取视频流信息".to_string()))?;

    let format = json.get("format")
        .ok_or_else(|| VsubError::Video("无法获取格式信息".to_string()))?;

    // 查找视频流
    let video_stream = streams.iter()
        .find(|s| s.get("codec_type").and_then(|c| c.as_str()) == Some("video"));

    // 查找音频流
    let audio_stream = streams.iter()
        .find(|s| s.get("codec_type").and_then(|c| c.as_str()) == Some("audio"));

    let duration = format.get("duration")
        .and_then(|d| d.as_str())
        .and_then(|s| s.parse::<f64>().ok())
        .or_else(|| video_stream.and_then(|s| s.get("duration").and_then(|d| d.as_str()).and_then(|s| s.parse::<f64>().ok())))
        .unwrap_or(0.0);

    let (width, height) = video_stream
        .map(|s| {
            let width = s.get("width").and_then(|w| w.as_u64()).unwrap_or(0) as u32;
            let height = s.get("height").and_then(|h| h.as_u64()).unwrap_or(0) as u32;
            (width, height)
        })
        .unwrap_or((0, 0));

    let fps = video_stream
        .and_then(|s| s.get("r_frame_rate").and_then(|r| r.as_str()))
        .and_then(|r| {
            let parts: Vec<&str> = r.split('/').collect();
            if parts.len() == 2 {
                let num = parts[0].parse::<f64>().ok()?;
                let den = parts[1].parse::<f64>().ok()?;
                Some(if den > 0.0 { num / den } else { 0.0 })
            } else {
                None
            }
        })
        .unwrap_or(0.0);

    let video_codec = video_stream
        .and_then(|s| s.get("codec_name").and_then(|c| c.as_str()).map(|s| s.to_string()));

    let audio_codec = audio_stream
        .and_then(|s| s.get("codec_name").and_then(|c| c.as_str()).map(|s| s.to_string()));

    debug!(
        "Video info: {}x{} @ {:.2}fps, duration: {:.2}s, audio: {}",
        width, height, fps, duration, audio_stream.is_some()
    );

    Ok(VideoInfo {
        path: path.to_path_buf(),
        duration,
        width,
        height,
        has_audio: audio_stream.is_some(),
        audio_codec,
        video_codec,
        fps,
    })
}

/// 验证视频文件
pub fn validate_video<P: AsRef<Path>>(path: P) -> Result<(), VsubError> {
    let path = path.as_ref();

    if !path.exists() {
        return Err(VsubError::Video(format!(
            "文件不存在: {}",
            path.display()
        )));
    }

    if !path.is_file() {
        return Err(VsubError::Video(format!(
            "不是文件: {}",
            path.display()
        )));
    }

    // 检查扩展名
    let ext = path.extension()
        .and_then(|e| e.to_str())
        .map(|e| e.to_lowercase())
        .unwrap_or_default();

    let supported = ["mp4", "mkv", "avi", "mov", "wmv", "flv", "webm", "m4v", "ts", "m2ts"];
    if !supported.contains(&ext.as_str()) {
        debug!("Unknown video extension: {}", ext);
    }

    // 使用 ffprobe 验证
    probe_video(path)?;

    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_check_ffmpeg() {
        // 在 CI 环境中可能失败
        let _ = check_ffmpeg();
    }
}
